File: 03-Artificial-Neural-Networks/csharp-neural-network/Network.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CSharp_Neural_Network
{
    class Network
    {
        readonly InputLayer inputLayer;
        readonly HiddenLayer[] hiddenLayers;
        readonly OutputLayer outputLayer;
        readonly Activation activation;

        public int InputSize { get; private set; }
        public int[] HiddenSizes { get; private set; }
        public int OutputSize { get; private set; }

        public Network(int inputSize, int[] hiddenSizes, int outputSize, double learnRate)
            : this(Activation.Sigmoid, new TrainingInfo(learnRate), inputSize, hiddenSizes, outputSize)
        { }

        public Network(Activation activation, TrainingInfo trainInfo, int inputSize, int[] hiddenSizes, int outputSize)
        {
            this.InputSize = inputSize;
            this.HiddenSizes = hiddenSizes;
            this.OutputSize = outputSize;

            this.activation = activation;
            this.inputLayer = new InputLayer(inputSize);
            this.hiddenLayers = hiddenSizes
                .Select(size => new HiddenLayer(activation, trainInfo, size))
                .ToArray();
            this.outputLayer = new OutputLayer(activation, trainInfo, outputSize);
            ConnectLayers();
        }

        void ConnectLayers()
        {
            if (!hiddenLayers.Any())
            {
                ConnectAdjacentLayers(inputLayer, outputLayer);
            }
            else
            {
                var lastHiddenIndex = hiddenLayers.Length - 1;
                ConnectAdjacentLayers(inputLayer, hiddenLayers[0]);
                for (int i = 0; i < lastHiddenIndex; i++)
                    ConnectAdjacentLayers(hiddenLayers[i], hiddenLayers[i + 1]);
                ConnectAdjacentLayers(hiddenLayers[lastHiddenIndex], outputLayer);
            }
        }

        void ConnectAdjacentLayers(Layer fromLayer, Layer toLayer)
        {
            foreach (var fromNeuron in fromLayer.Neurons)
                foreach (var toNeuron in toLayer.Neurons)
                    new Connection(fromNeuron, toNeuron, activation.GetRandomWeight());
        }

        public double[] FeedForward(double[] input)
        {
            inputLayer.SetInputValues(input);
            foreach (var hiddenLayer in hiddenLayers)
                hiddenLayer.FeedForward();
            outputLayer.FeedForward();
            return outputLayer.Values;
        }

        public void PropagateBack(double[] target)
        {
            outputLayer.SetTargetValues(target);
            outputLayer.PropagateBack();
            foreach (var hiddenLayer in hiddenLayers.Reverse())
                hiddenLayer.PropagateBack();
        }

    }
}

File: 03-Artificial-Neural-Networks/fsharp-neural-network2/Network.fs
module Network

open Layer
open Activation
open TrainingInfo

type Network = 
   {InputLayer : Layer
    NonInputLayers : Layer[] // hiden layers + output layer
    HiddenLayers : Layer[]
    OutputLayer : Layer
    Activation : Activation
    TrainingInfo : TrainingInfo}

let createNetwork inputSize hiddenSizes outputSize learnRate = 
    let activation = Activation.sigmoidActivation
    let inputLayer = createLayer inputSize 0 activation.GetRandomWeight 
    let allLayers = 
        (inputLayer, hiddenSizes @ [outputSize])
        ||> List.scan (fun previousLayer size ->
            createLayer size (previousLayer.Neurons.Length + 1) activation.GetRandomWeight )
    let nonInputLayers = allLayers.Tail |> List.toArray
    let hiddenLayers = nonInputLayers.[0 .. (nonInputLayers.Length-2)]
    {   InputLayer = inputLayer
        NonInputLayers = nonInputLayers
        HiddenLayers = hiddenLayers
        OutputLayer = nonInputLayers.[nonInputLayers.Length - 1]
        Activation = activation
        TrainingInfo = {LearnRate = learnRate} }

let getLayerValues layer =
    seq{ for neuron in layer.Neurons do yield neuron.Value}

let getLayerValuesAndBias layer = 
    seq{ 
        yield! getLayerValues layer 
        yield 1.}
    |> Seq.toArray


let feedForward net input =
    (net.InputLayer.Neurons, input) 
        ||> Seq.iter2 (fun n i -> n.Value <- i)
    let outputLayer =
        (net.InputLayer, net.NonInputLayers)
        ||> Array.fold (fun previousLayer layer ->
            let previousLayerValues = getLayerValuesAndBias previousLayer
            layer.Neurons 
            |> Array.iter (fun neuron ->
                let sumOfIncoming = 
                    (0., neuron.InboundWeights, previousLayerValues) 
                    |||> Array.fold2 (fun sum weight pv -> sum + weight * pv)
                neuron.Value <- net.Activation.CalcValue sumOfIncoming)
            layer)
    getLayerValues outputLayer |> Seq.toArray


let propagateBack net target =
    let adjustWeightsBetweenLayers inputLayer layer =
        let inputLayerValues = getLayerValuesAndBias inputLayer
        layer.Neurons
        |> Array.iter (fun neuron ->
            inputLayerValues |> Array.iteri (fun index inputValue->
                let weightDelta = net.TrainingInfo.LearnRate * neuron.Error * inputValue
                neuron.InboundWeights.[index] <- neuron.InboundWeights.[index] + weightDelta))

    // set output layer errors
    (target, net.OutputLayer.Neurons)
    ||> Array.iter2 (fun targetValue neuron -> 
        let valueDelta = targetValue - neuron.Value
        neuron.Error <- valueDelta * net.Activation.CalcDerivative neuron.Value)
         
    // foldback updating neuron errors and adjusting weights
    let layerAfterInput = 
        (net.HiddenLayers, net.OutputLayer)
        ||> Array.foldBack(fun layer outLayer ->
            layer.Neurons
            |> Array.iteri(fun index neuron -> 
                let valueDelta =  
                    (0., outLayer.Neurons)
                    ||> Array.fold (fun error outNeuron ->
                        error + (outNeuron.InboundWeights.[index] * outNeuron.Error))
                neuron.Error <- valueDelta * net.Activation.CalcDerivative neuron.Value)
            adjustWeightsBetweenLayers layer outLayer
            layer)
    
    adjustWeightsBetweenLayers net.InputLayer layerAfterInput
    
   
